assets/src/edit-story/components/canvas/gridview/index.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * External dependencies
+ */
+import styled from 'styled-components';
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import { useCallback, useState } from '@wordpress/element';
+import { __, sprintf } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { useStory } from '../../../app/story';
+import DropZone from '../../dropzone';
+import RectangleIcon from './rectangle.svg';
+
+const PAGE_WIDTH = 90;
+const PAGE_HEIGHT = 160;
+
+const Wrapper = styled.div`
+	position: relative;
+	display: grid;
+	grid-template-columns: ${ ( { scale } ) => `repeat(auto-fit, minmax(${ scale * PAGE_WIDTH }px, max-content))` };
+	grid-gap: 20px;
+	justify-content: center;
+	justify-items: center;
+    align-items: center;
+`;
+
+const Page = styled.button`
+	padding: 0;
+	margin: 0;
+	border: 3px solid ${ ( { isActive, theme } ) => isActive ? theme.colors.selection : theme.colors.bg.v1 };
+	height: ${ ( { scale } ) => `${ scale * PAGE_HEIGHT }px` };
+	width: ${ ( { scale } ) => `${ scale * PAGE_WIDTH }px` };
+	background-color: ${ ( { isActive, theme } ) => isActive ? theme.colors.fg.v1 : theme.colors.mg.v1 };
+	transition: width .2s ease, height .2s ease;
+`;
+
+const RangeInputWrapper = styled.div`
+	display: flex;
+	width: 400px;
+	margin: 0 auto;
+`;
+
+const RangeInput = styled.input.attrs( () => ( {
+	type: 'range',
+} ) )`
+	display: block;
+	width: 360px;
+	margin: 0 auto;
+`;
+
+function RangeControl( { value, onChange } ) {
+	return (
+		<RangeInputWrapper>
+			<RectangleIcon />
+			<RangeInput
+				min="1"
+				max="3"
+				steps="1"
+				value={ value }
+				onChange={ ( evt ) => onChange( Number( evt.target.value ) ) }
+			/>
+			<RectangleIcon />
+		</RangeInputWrapper>
+	);
+}
+
+RangeControl.propTypes = {
+	value: PropTypes.number.isRequired,
+	onChange: PropTypes.func.isRequired,
+};
+
+// TODO: Make drag & drop part DRY.
+function GridView() {
+	const { state: { pages, currentPageIndex }, actions: { setCurrentPage, arrangePage } } = useStory();
+	const [ zoomLevel, setZoomLevel ] = useState( 2 );
+
+	const getArrangeIndex = ( sourceIndex, dstIndex, position ) => {
```

Comment:
Thinking about creating a new `DraggablePage` component to fix all this duplicate code around drag & drop here.